File: data/EXP01/python/2300/2300.output_3.py
# @lc app=leetcode id=2300 lang=python3
from typing import List

class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        pairs = [0] * len(spells) # Initialize the result list with zeros
        
        potions.sort() # Sort the potions array in ascending order
        
        for i, spell in enumerate(spells):
            left, right = 0, len(potions) - 1 # Pointers for left and right ends of potions array
            
            while left <= right:
                mid = (left + right) // 2
                if spell * potions[mid] >= success:
                    pairs[i] += len(potions) - mid
                    right = mid - 1
                else:
                    left = mid + 1
        
        return pairs

File: data/EXP01/golang_no_system_var/2300/2300.output_3.go
func successfulPairs(spells []int, potions []int, success int64) []int {
    sort.Ints(potions)
    pairs := make([]int, len(spells))
    
    for i := 0; i < len(spells); i++ {
        spell := int64(spells[i])
        count := binarySearch(potions, success/spell) - 1
        if count < 0 {
            count = 0
        }
        pairs[i] = count
    }
    
    return pairs
}

func binarySearch(arr []int, target int64) int {
    left := 0
    right := len(arr)
    
    for left < right {
        mid := left + (right-left)/2
        if int64(arr[mid]) < target {
            left = mid + 1
        } else {
            right = mid
        }
    }
    
    return left
}
